proof/refine/ARM/KHeap_R.thy

```diff
@@ -1375,38 +1375,113 @@ lemma set_reply_corres: (* for reply update that doesn't touch the reply stack *
         simp add: typ_at'_def ko_wp_at'_def obj_at'_def project_inject opt_map_def sc_of_def)
   qed
 
-lemma setSchedContext_no_stack_update_corres:
-  "\<lbrakk>sc_relation sc n sc' \<longrightarrow> sc_relation (f sc) n (f' sc');
-     \<forall>sc. sc_replies sc = sc_replies (f sc); \<forall>sc'. objBits sc' = objBits (f' sc');
-     scReply sc' = scReply (f' sc')\<rbrakk>
-    \<Longrightarrow> corres dc
+lemma setReply_corres: (* for reply updates on replies not in fst ` replies_with_sc *)
```

Comment:
Some parts of this look a bit similar to some of the `setSchedContext` corres rules that @mktnk3  has. Is there any way to refactor things to reduce duplication?

Reply:
I spent a few hours on this, but it's a lot of work to unify things. The reason they look similar is that they're both based on the generic `set_object` corres proofs. But in each case because the state-relation is slightly non-trivial on these particular kernel objects, the generic argument doesn't quite work. In the future it would be nice to unify these somehow, but it would require quite a lot of time to get it working.